File: cmd/protocol.go
/*
 * Copyright (C) 2024 Delusoire
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

package cmd

import (
	"fmt"
	"net/url"
	"os/exec"
	"runtime"
	"spicetify/module"
	"strings"

	e "spicetify/errors"

	"github.com/spf13/cobra"
)

var protocolCmd = &cobra.Command{
	Use:   "protocol [uri]",
	Short: "Internal protocol handler",
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		res, err := HandleProtocol(args[0])
		if res != "" {
			open("spotify:app:rpc:" + res)
		}
		if err != nil {
			fmt.Println(err)
		}
	},
}

func HandleProtocol(uri string) (string, error) {
	u, err := url.Parse(uri)
	if err != nil || u.Scheme != "spicetify" {
		return "", err
	}
	uuid, action, _ := strings.Cut(u.Opaque, ":")
	response := u.Scheme + ":" + uuid + ":"
	arguments := u.Query()
	err = hp(action, arguments)
	if err == nil {
		response += "1"
	} else {
		response += "0"
	}
	if uuid == "0" {
		response = ""
	}
	return response, err
}

func hp(action string, arguments url.Values) error {
	switch action {
	case "add":
		_artifacts := arguments["artifacts"]

		identifier := module.NewStoreIdentifier(arguments.Get("id"))
		artifacts := make([]module.ArtifactURL, len(_artifacts))
		for i, a := range _artifacts {
			artifacts[i] = module.ArtifactURL(a)
		}
		checksum := arguments.Get("checksum")

		return module.AddStoreInVault(identifier, &module.Store{
			Installed: false,
			Artifacts: artifacts,
			Checksum:  checksum,
		})

	case "install":
		identifier := module.NewStoreIdentifier(arguments.Get("id"))
		return module.InstallModule(identifier)

	case "enable":
		identifier := module.NewStoreIdentifier(arguments.Get("id"))
		return module.EnableModuleInVault(identifier)

	case "delete":
		identifier := module.NewStoreIdentifier(arguments.Get("id"))
		return module.DeleteModule(identifier)

	case "remove":
		identifier := module.NewStoreIdentifier(arguments.Get("id"))
		return module.RemoveStoreInVault(identifier)

	}
	return e.ErrUnsupportedOperation
}

func init() {
	rootCmd.AddCommand(protocolCmd)
}

func open(url string) error {
	var cmd string
	var args []string

	switch runtime.GOOS {
	case "windows":
		cmd = "cmd"
		args = []string{"/c", "start"}
	case "darwin":
		cmd = "open"
	default:
		cmd = "xdg-open"
	}
	args = append(args, url)
	return exec.Command(cmd, args...).Start()
}

File: src/commands/protocol.rs
// TODO: register spicetify:// URI
// Go's README.md:69 lists "register uri scheme and daemon/task" as a TODO. On Windows this
// means writing registry keys (HKEY_CLASSES_ROOT\spicetify\shell\open\command) pointing to the
// spicetify binary. On macOS it's a LaunchServices/CFBundleURLSchemes entry. On Linux it's a
// .desktop file with x-scheme-handler/spicetify MIME type. Without this, spicetify:// URIs from
// the marketplace UI won't trigger the protocol handler automatically.

use std::process::Command;

use anyhow::{Result, bail};
use url::Url;

use crate::{
    config::AppContext, module::{self, ModulePaths, Store}
};

pub fn run(ctx: &AppContext, uri: &str) -> Result<()> {
    let response = handle(ctx, uri)?;
    if !response.is_empty() {
        let outbound = format!("spotify:app:rpc:{}", response);
        launch_uri(&outbound)?;
    }
    Ok(())
}

pub fn handle(ctx: &AppContext, uri: &str) -> Result<String> {
    let u = Url::parse(uri)?;
    if u.scheme() != "spicetify" {
        bail!("unsupported scheme")
    }

    let opaque = u.path();
    let mut parts = opaque.split(':');
    let uuid = parts.next().unwrap_or_default();
    let action = parts.next().unwrap_or_default();

    let prefix = format!("spicetify:{}:", uuid);
    let result = perform(ctx, action, &u);

    let mut response = prefix;
    response.push(if result.is_ok() { '1' } else { '0' });
    result?;

    if uuid == "0" {
        Ok(String::new())
    } else {
        Ok(response)
    }
}

fn perform(ctx: &AppContext, action: &str, uri: &Url) -> Result<()> {
    let paths = ModulePaths::from_config_root(&ctx.config_root);
    let query: Vec<_> = uri.query_pairs().collect();

    let get = |key: &str| -> Option<String> {
        query
            .iter()
            .find(|(k, _)| k == key)
            .map(|(_, v)| v.to_string())
    };
    let get_all = |key: &str| -> Vec<String> {
        query
            .iter()
            .filter(|(k, _)| k == key)
            .map(|(_, v)| v.to_string())
            .collect()
    };

    match action {
        "add" | "fast-install" | "fast-enable" => {
            let id = module::vault::StoreIdentifier::parse(&get("id").unwrap_or_default())?;
            let artifacts = get_all("artifacts");
            let checksum = get("checksum").unwrap_or_default();

            module::add_store(
                &paths,
                &id,
                Store {
                    installed: false,
                    artifacts,
                    checksum,
                },
            )?;

            if action == "add" {
                return Ok(());
            }
            module::install(&paths, &id)?;
            if action == "fast-install" {
                return Ok(());
            }
            module::enable(&paths, &id)
        }
        "install" => {
            let id = module::vault::StoreIdentifier::parse(&get("id").unwrap_or_default())?;
            module::install(&paths, &id)
        }
        "enable" => {
            let id = module::parse_enable_id(&get("id").unwrap_or_default())?;
            module::enable(&paths, &id)
        }
        "delete" => {
            let id = module::vault::StoreIdentifier::parse(&get("id").unwrap_or_default())?;
            module::delete(&paths, &id)
        }
        "remove" => {
            let id = module::vault::StoreIdentifier::parse(&get("id").unwrap_or_default())?;
            module::remove_store(&paths, &id)
        }
        "fast-delete" | "fast-remove" => {
            let id = module::vault::StoreIdentifier::parse(&get("id").unwrap_or_default())?;
            let disable_id = module::vault::StoreIdentifier {
                module_identifier: id.module_identifier.clone(),
                version: String::new(),
            };
            module::enable(&paths, &disable_id)?;
            module::delete(&paths, &id)?;
            if action == "fast-remove" {
                module::remove_store(&paths, &id)?;
            }
            Ok(())
        }
        _ => bail!("this operation is not supported"),
    }
}

fn launch_uri(uri: &str) -> Result<()> {
    let result = if cfg!(target_os = "windows") {
        Command::new("cmd").args(["/c", "start", uri]).spawn()
    } else if cfg!(target_os = "macos") {
        Command::new("open").arg(uri).spawn()
    } else {
        Command::new("xdg-open").arg(uri).spawn()
    };

    result
        .map(|_| ())
        .map_err(|e| anyhow::anyhow!("failed to open URI {uri}: {e}"))
}
